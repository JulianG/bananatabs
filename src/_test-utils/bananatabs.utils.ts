// tslint:disable no-unused-variable
import * as React from 'react';
import {
  render,
  queryAllByAttribute,
  queryByAltText,
  getByTestId
} from 'react-testing-library';
import { stringToSession } from '../serialisation/MarkdownSerialisation';
import { FakePromisingChromeAPI } from '../chrome-api/FakePromisingChromeAPI';
import { createFakeBananaContext } from '../context/BananaContext';
import { BananaTabs } from '../BananaTabs';
import { wait } from '.';

export { wait };

export async function renderBananaTabs(
  live: string,
  stored: string | null = null
) {
  stored = stored !== null ? stored : live;
  const context = getBananaContext(live, stored);
  const fchrome = context.chromeAPI as FakePromisingChromeAPI;

  const { container, ...renderResult } = render(
    React.createElement(BananaTabs, { context: context })
  );
  await wait();
  const provider = context.sessionProvider;

  return {
    container,
    fchrome,
    provider,
    ...renderResult,
    ...createFunctions(container)
  };
}

function createFunctions(container: HTMLElement) {
  return {
    getWindowGroups: () => {
      return getWindowGroups(container);
    },
    getTab: (windowIndex: number, tabIndex: number) => {
      return getTabsInWindow(getWindowGroups(container)[windowIndex])[tabIndex];
    },
    getTabVisibilityToggle: (windowIndex: number, tabIndex: number) => {
      const tab = getTabsInWindow(getWindowGroups(container)[windowIndex])[
        tabIndex
      ];
      return getByTestId(tab, 'visibility-toggle');
    },
    getWindowVisibilityToggle: (windowIndex: number) => {
      const win = getWindowGroups(container)[windowIndex];
      return getByTestId(win, 'visibility-toggle');
    },
    getWindowDisclosureButton: (windowIndex: number) => {
      const win = getWindowGroups(container)[windowIndex];
      const expanded = queryByAltText(win, 'window-disclosure-expanded');
      const collapsed = queryByAltText(win, 'window-disclosure-collapsed');
      if (!expanded && !collapsed) {
        throw new Error(
          'Could not find alt="window-disclosure-expanded" nor alt="window-disclosure-collapsed"'
        );
      }
      return expanded || collapsed!;
    },
    isWindowExpanded: (windowIndex: number) => {
      const win = getWindowGroups(container)[windowIndex];
      const expanded = queryByAltText(win, 'window-disclosure-expanded');
      const collapsed = queryByAltText(win, 'window-disclosure-collapsed');
      if (!expanded && !collapsed) {
        throw new Error(
          'Could not find alt="window-disclosure-expanded" nor alt="window-disclosure-collapsed"'
        );
      }
      return !!expanded;
    }
  };
}

function getBananaContext(live: string, stored: string) {
  const liveSession = stringToSession(live);
  const storedSession = stringToSession(stored);
  const fake = { live: liveSession, stored: storedSession };
  return createFakeBananaContext(fake);
}

export function getWindowGroups(container: HTMLElement) {
  return queryAllByAttribute('id', container, 'window-group');
}

export function getTabsInWindow(windowGroup: HTMLElement) {
  return queryAllByAttribute('id', windowGroup, 'tab');
}

export function getTabsVisibilities(tabs: HTMLElement[]): boolean[] {
  return tabs.map(t => {
    const visible = queryByAltText(t, 'tab-visible');
    // const hidden = queryByAltText(t, 'tab-hidden');
    return visible !== null;
  });
}

export function getWindowsVisibilities(windows: HTMLElement[]): boolean[] {
  return windows.map(w => {
    const visible = queryByAltText(w, 'win-visible');
    return visible !== null;
  });
}
