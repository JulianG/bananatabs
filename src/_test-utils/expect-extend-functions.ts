import * as BT from '../model/CoreTypes';

interface ChromeEvent {
  addListener: Function;
  removeListener: Function;
}

type Callback = jest.Mock<{}>;

interface Expectation {
  event: ChromeEvent;
  times: number;
}

export type AllCallbacks = ChromeEventWithCallback[];

interface ChromeEventWithCallback {
  id: string;
  event: ChromeEvent;
  callback: Callback;
}

expect.extend({ toHaveBeenCalledLike: confirmExpectations });

function confirmExpectations(
  allCallbacks: AllCallbacks,
  expectations: Expectation[]
) {
  for (let i = 0; i < allCallbacks.length; i++) {
    const pair = allCallbacks[i];

    const expectation = expectations.find(ex => ex.event === pair.event);
    const times = expectation ? expectation.times : 0;

    if (pair.callback.mock.calls.length !== times) {
      return {
        message: () => {
          return (
            `expected event '${pair.id}' \nto have been triggered: ` +
            `  ${times} times, but` +
            `\ninstead it was triggered:` +
            ` ${pair.callback.mock.calls.length} times.`
          );
        },
        pass: false,
      };
    }
  }

  return {
    message: () => `all the expectations where correct`,
    pass: true,
  };
}

// this extension is here because `toHaveBeenNthCalledWith` seems to have disappeared
expect.extend({
  toHaveBeenCalledNthTimeWith: (
    fn: jest.Mock<{}>,
    nthCall: number,
    expected: {}
  ) => {
    if (fn.mock.calls.length < nthCall) {
      return {
        message: () => {
          return `expected function to have been called at least ${nthCall} times,
but it was instead called ${fn.mock.calls.length} times`;
        },
        pass: false,
      };
    }

    const args = fn.mock.calls[nthCall - 1];

    if (JSON.stringify(args) !== JSON.stringify(expected)) {
      return {
        message: () => {
          return `expected function to have been called with ${JSON.stringify(
            expected
          )} on the ${nthCall} time,
but instead it was called with ${JSON.stringify(args)}`;
        },
        pass: false,
      };
    } else {
      return {
        message: () => `all the expectations where correct`,
        pass: true,
      };
    }
  },
});

// ///

expect.extend({
  toBeEquivalentToBTWindow: (
    actual: chrome.windows.Window,
    expected: BT.Window
  ) => {
    const visiblePass = expected.visible;
    const focusPass = actual.focused === expected.focused;
    const actualURLs = (actual.tabs || []).map(t => t.url);
    const expectedURLs = expected.tabs.filter(t => t.visible).map(t => t.url);
    const urlsLengthPass = actualURLs.length === expectedURLs.length;
    const urlsPass = actualURLs.every((url, i) => url === expectedURLs[i]);

    if (focusPass && urlsLengthPass && urlsPass && visiblePass) {
      return {
        message: () =>
          `expected windows not to be equivalent, but they have the same 'focused' value and same list of tabs.`,
        pass: true,
      };
    } else {
      let message = '';

      if (!visiblePass) {
        message = `the expected window must be visible`;
      } else if (!focusPass) {
        message = `expected focused to be ${
          expected.focused
        }, but instead it was ${actual.focused}`;
      } else if (!urlsLengthPass || !urlsPass) {
        message = `expected URLs to be ${JSON.stringify(expectedURLs, null, 2)},
but instead it was ${JSON.stringify(actualURLs, null, 2)}`;
      }
      return { message: () => message, pass: false };
    }
  },
});
