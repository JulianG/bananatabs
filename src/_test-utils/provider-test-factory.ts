import * as BT from '../model/CoreTypes';
import ChromeBrowserController from '../chrome/ChromeBrowserController';
import LiveSessionMerger, { DefaultLiveSessionMerger } from '../model/mergers/LiveSessionMerger';
import RAMSessionPersistence from '../utils/RAMSessionPersistence';
import DefaultSessionProvider from '../model/DefaultSessionProvider';
import { initialiseFakeChromeAPI } from '../utils/initialise-fake-chrome-api';
import FakePromisingChromeAPI from 'chrome-api/FakePromisingChromeAPI';
import { parseSessionString } from '../utils/session-string-parser';

export async function createProvider(sessionString: string) {
	const session = parseSessionString(sessionString);
	const fchrome = initialiseFakeChromeAPI(session);
	return { fchrome, ...createProviderWFC(fchrome, session) };
}

export async function createIniatilisedProvider(sessionString: string) {
	const session = parseSessionString(sessionString);
	const fchrome = initialiseFakeChromeAPI(session);
	return createInitialisedProviderWFC(fchrome, session);
}

async function createInitialisedProviderWFC(fchrome: FakePromisingChromeAPI, session: BT.Session) {
	const { provider, onSessionChanged, browserController } = createProviderWFC(fchrome, session);
	provider.onSessionChanged = onSessionChanged;
	await provider.initialiseSession('jest');
	onSessionChanged.mockReset();
	return { provider, onSessionChanged, fchrome, browserController };
}

function createProviderWFC(fchrome: FakePromisingChromeAPI, session: BT.Session) {
	const browserController = new ChromeBrowserController(fchrome);
	const merger: LiveSessionMerger = new DefaultLiveSessionMerger();
	const persistence = new RAMSessionPersistence(session);
	const provider = new DefaultSessionProvider(browserController, merger, persistence);
	const onSessionChanged = jest.fn();
	provider.onSessionChanged = onSessionChanged;
	return { provider, onSessionChanged, browserController };
}