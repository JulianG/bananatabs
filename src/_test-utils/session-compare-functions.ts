import * as BT from '../model/core/CoreTypes';

export function compareSessions(s0: BT.Session, s1: BT.Session): boolean {
  const sw0 = s0.windows;
  const sw1 = s1.windows;
  const rsp =
    sw0.length === sw1.length &&
    sw0.every((w0, i) => {
      const w1 = sw1[i];
      return compareWindows(w0, w1);
    });
  if (!rsp) {
    throw `Error: compareSessions: sessions don't match.\n
		s0.windows.length: ${s0.windows.length}\n
		s1.windows.length: ${s1.windows.length}`;
  }
  return rsp;
}

function compareWindows(w0: BT.Window, w1: BT.Window): boolean {
  if (w0.title !== w1.title) {
    throw `Error: compareWindows: window titles don't match:\n
		${w0.title} !== ${w1.title}\n`;
  }
  if (w0.visible !== w1.visible) {
    throw `Error: compareWindows: window visibilities don't match:\n
		${w0.visible} !== ${w1.visible}\n`;
  }
  if (compareTabLists(w0.tabs, w1.tabs) === false) {
    throw `Error: compareWindows: the list of tabs on each window don't match:\n
		${w0.tabs} !== ${w1.tabs}\n`;
  }
  return true;
}

function compareTabLists(tl0: ReadonlyArray<BT.Tab>, tl1: ReadonlyArray<BT.Tab>): boolean {
  const rsp =
    tl0.length === tl1.length && tl0.every((t0, i) => compareTab(t0, tl1[i]));
  if (!rsp) {
    const l0: string = JSON.stringify(tl0, null, 2);
    const l1: string = JSON.stringify(tl1, null, 2);
    throw `Error: compareTabLists:\n${l0}\n\n${l1}`;
  }
  return rsp;
}

function compareTab(t0: BT.Tab, t1: BT.Tab): boolean {
  if (t0.url !== t1.url) {
    const s0 = serialiseItem(t0);
    const s1 = serialiseItem(t1);
    throw `Error: compareTab: non-matching urls in the following tabs:
						\n${s0}\n${s1}
						\n${t0.url} != ${t1.url}`;
  }
  if (t0.visible !== t1.visible) {
    const s0 = serialiseItem(t0);
    const s1 = serialiseItem(t1);
    throw `Error: compareTab: non-matching 'visible' in the following tabs:
						\n${s0}\n${s1}
						\n${t0.visible} != ${t1.visible}`;
  }

  return true;
}

function sortKeys(unordered: Object) {
  const ordered = {};
  Object.keys(unordered)
    .sort()
    .forEach(function(key: string) {
      ordered[key] = unordered[key];
    });
  return ordered;
}

function serialiseItem(item: Object) {
  return JSON.stringify(sortKeys({ ...item, id: 'xxx' }), null, 2);
}
