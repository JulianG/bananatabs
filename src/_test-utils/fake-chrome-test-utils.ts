import { FakePromisingChromeAPI } from '../chrome-api/FakePromisingChromeAPI';
import { AllCallbacks } from './expect-extend-functions';

export function getAllCallbacks(fchrome: FakePromisingChromeAPI) {
  const allCallbacks: AllCallbacks = [
    {
      id: 'chrome.windows.onCreated',
      event: fchrome.windows.onCreated,
      callback: jest.fn()
    },
    {
      id: 'chrome.windows.onFocusChanged',
      event: fchrome.windows.onFocusChanged,
      callback: jest.fn()
    },
    {
      id: 'chrome.windows.onRemoved',
      event: fchrome.windows.onRemoved,
      callback: jest.fn()
    },
    {
      id: 'chrome.tabs.onCreated',
      event: fchrome.tabs.onCreated,
      callback: jest.fn()
    },
    {
      id: 'chrome.tabs.onActivated',
      event: fchrome.tabs.onActivated,
      callback: jest.fn()
    },
    {
      id: 'chrome.tabs.onAttached',
      event: fchrome.tabs.onAttached,
      callback: jest.fn()
    },
    {
      id: 'chrome.tabs.onMoved',
      event: fchrome.tabs.onMoved,
      callback: jest.fn()
    },
    {
      id: 'chrome.tabs.onRemoved',
      event: fchrome.tabs.onRemoved,
      callback: jest.fn()
    },
    {
      id: 'chrome.tabs.onUpdated',
      event: fchrome.tabs.onUpdated,
      callback: jest.fn()
    },
    {
      id: 'chrome.tabs.onHighlighted',
      event: fchrome.tabs.onHighlighted,
      callback: jest.fn()
    }
  ];

  hookAllCallbacks(allCallbacks);

  return allCallbacks;
}

function hookAllCallbacks(allCallbacks: AllCallbacks) {
  allCallbacks.forEach(entry => {
    entry.callback.mockClear();
    entry.callback.mockReset();
    entry.event.addListener(entry.callback);
  });
}

export function resetAllCallbacks(allCallbacks: AllCallbacks) {
  allCallbacks.forEach(entry => {
    entry.callback.mockClear();
    entry.callback.mockReset();
  });
}

export function countCalls(allCallbacks: AllCallbacks) {
  console.log(
    allCallbacks.map(
      entry => `${entry.id}: ${entry.callback.mock.calls.length}`
    )
  );
}
