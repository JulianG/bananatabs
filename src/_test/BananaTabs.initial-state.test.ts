import 'react-testing-library/cleanup-after-each';

import {
  getWindowGroups,
  getTabsVisibilities,
  getTabsInWindow,
  renderBananaTabs,
} from '../_test-utils/bananatabs.utils';

import { wait } from '../_test-utils';

describe('BananaTabs Tests: Initial State', async () => {
  //
  test('initial UI snapshot', async () => {
    //
    // given an initial state with 2 windows
    const windows = `
			window 1:
			* http://tab-1.1/
			* http://tab-1.2/
			* http://tab-1.3/

			window 2:
			* http://tab-2.1/
			* http://tab-2.2/
			* http://tab-2.3/
    `;

    // when the app starts and is rendered
    const { container } = await renderBananaTabs(windows);
    await wait(1);

    // expect the rendered output to match the snapshot
    expect(container).toMatchSnapshot();
  });

  test('identical initial states', async () => {
    //
    // given an initial state with
    // 2 live windows
    // and 2 stored windows
    // which are identical

    const live = `
      window 1:
      * http://tab-1.1/
      * http://tab-1.2/
      * http://tab-1.3/

      window 2:
      * http://tab-2.1/
      * http://tab-2.2/
      * http://tab-2.3/
    `;
    const stored = live;

    // when the app starts and is rendered
    const { container, fchrome } = await renderBananaTabs(live, stored);
    await wait(1);

    // expect 2 window groups
    // because the live windows were matched to stored windows
    const windowGroups = getWindowGroups(container);
    expect(windowGroups).toHaveLength(2);
    expect(fchrome.fakeWindows).toHaveLength(2);

    const [window1, window2] = windowGroups;

    // also expect 3 visible tabs in the first window
    expect(getTabsVisibilities(getTabsInWindow(window1))).toMatchObject([
      true,
      true,
      true,
    ]);

    // also expect 3 visible tabs in the second window
    expect(getTabsVisibilities(getTabsInWindow(window2))).toMatchObject([
      true,
      true,
      true,
    ]);
    // as described in the initialisation strings
  });
});
