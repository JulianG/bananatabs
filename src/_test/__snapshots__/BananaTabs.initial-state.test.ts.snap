// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BananaTabs Tests: Initial State initial UI snapshot 1`] = `
<div>
  <div>
    <div>
      <h3>
        <img
          class="app-icon"
          src="/icons/app-icon.png"
        />
         
        <span>
          Banana Tabs!
        </span>
         
        <div
          class="credits"
          style="display: inline;"
        >
          BETA
        </div>
      </h3>
    </div>
    <div>
      <div
        class="session dl-list"
        style="display: block;"
      >
        <div
          class="dl-item   "
        >
          <div
            class="window-group   visible"
            id="window-group"
          >
            <div
              class="item-row"
            >
              <img
                class="tool icon "
                id="disclosure"
                src="[object Object]"
                title="Collapse"
              />
              <img
                class="tool icon"
                id="visibility-visible"
                src="[object Object]"
                title="Hide Window"
              />
              <span
                class="window-title"
              >
                window 1
                 
                <span />
                 
                <span />
              </span>
            </div>
            <div
              class="item-row tab   visible"
              id="tab"
            >
              <img
                alt="tab-visible"
                class="left-most tool icon"
                data-testid="visibility-toggle"
                id="tab-visibility-visible"
                src="[object Object]"
                title="Hide Tab"
              />
              <img
                class="icon "
                src="[object Object]"
              />
              <span
                class="tab-title"
              >
                http://tab-1.1/
              </span>
            </div>
            <div
              class="item-row tab   visible"
              id="tab"
            >
              <img
                alt="tab-visible"
                class="left-most tool icon"
                data-testid="visibility-toggle"
                id="tab-visibility-visible"
                src="[object Object]"
                title="Hide Tab"
              />
              <img
                class="icon "
                src="[object Object]"
              />
              <span
                class="tab-title"
              >
                http://tab-1.2/
              </span>
            </div>
            <div
              class="item-row tab   visible"
              id="tab"
            >
              <img
                alt="tab-visible"
                class="left-most tool icon"
                data-testid="visibility-toggle"
                id="tab-visibility-visible"
                src="[object Object]"
                title="Hide Tab"
              />
              <img
                class="icon "
                src="[object Object]"
              />
              <span
                class="tab-title"
              >
                http://tab-1.3/
              </span>
            </div>
          </div>
        </div>
        <div
          class="dl-item   "
        >
          <div
            class="window-group   visible"
            id="window-group"
          >
            <div
              class="item-row"
            >
              <img
                class="tool icon "
                id="disclosure"
                src="[object Object]"
                title="Collapse"
              />
              <img
                class="tool icon"
                id="visibility-visible"
                src="[object Object]"
                title="Hide Window"
              />
              <span
                class="window-title"
              >
                window 2
                 
                <span />
                 
                <span />
              </span>
            </div>
            <div
              class="item-row tab   visible"
              id="tab"
            >
              <img
                alt="tab-visible"
                class="left-most tool icon"
                data-testid="visibility-toggle"
                id="tab-visibility-visible"
                src="[object Object]"
                title="Hide Tab"
              />
              <img
                class="icon "
                src="[object Object]"
              />
              <span
                class="tab-title"
              >
                http://tab-2.1/
              </span>
            </div>
            <div
              class="item-row tab   visible"
              id="tab"
            >
              <img
                alt="tab-visible"
                class="left-most tool icon"
                data-testid="visibility-toggle"
                id="tab-visibility-visible"
                src="[object Object]"
                title="Hide Tab"
              />
              <img
                class="icon "
                src="[object Object]"
              />
              <span
                class="tab-title"
              >
                http://tab-2.2/
              </span>
            </div>
            <div
              class="item-row tab   visible"
              id="tab"
            >
              <img
                alt="tab-visible"
                class="left-most tool icon"
                data-testid="visibility-toggle"
                id="tab-visibility-visible"
                src="[object Object]"
                title="Hide Tab"
              />
              <img
                class="icon "
                src="[object Object]"
              />
              <span
                class="tab-title"
              >
                http://tab-2.3/
              </span>
            </div>
          </div>
        </div>
        <div
          class="dl-item floating"
          style="pointer-events: none; position: absolute; left: 0px; top: 0px; width: 0px; height: 0px;"
        />
      </div>
      <div
        class="command-buttons"
      >
        <button
          class="ok"
        >
          Add links
        </button>
        <span />
        <button
          class="cancel"
        >
          Share all windows
        </button>
      </div>
    </div>
    <div
      class="credits"
    >
      <p>
        <strong>
          BananaTabs! 
          <a
            class="version"
            href="#"
            title=""
          >
            v
            0.27.3
          </a>
        </strong>
        <br />
        Developed by Julian Garamendy.
        <br />
        Icons designed by Dave Gandy from 
        <a
          href="https://www.flaticon.com/packs/font-awesome"
          target="_blank"
        >
          FlatIcon
        </a>
        . - Banana icon by Freepik from 
        <a
          href="https://www.flaticon.com/free-icon/banana_688828"
          target="_blank"
        >
          FlatIcon
        </a>
        .
      </p>
    </div>
  </div>
</div>
`;
